src/Xamarin.Android.Build.Tasks/Utilities/NdkTools/NdkTools.cs

```diff
@@ -0,0 +1,391 @@
+				// original tool name
+				if (Path.HasExtension (toolPath) && File.Exists (toolPath)) {
+					executablePath = toolPath;
+				} else {
+					string toolDir = Path.GetDirectoryName (toolPath);
+					executablePath = Path.Combine (toolDir, MonoAndroidHelper.GetExecutablePath (toolDir, Path.GetFileName (toolPath)));
+				}
+			} else if (File.Exists (toolPath)) {
+				executablePath = toolPath;
+			}
+
+			if (mustExist && String.IsNullOrEmpty (executablePath)) {
+				throw new InvalidOperationException ($"Required tool '{toolPath}' not found");
+			}
+
+			return executablePath;
+		}
+
+		protected virtual string GetToolName (NdkToolKind kind)
+		{
+			if (!NdkToolNames.TryGetValue (kind, out string? toolName) || String.IsNullOrEmpty (toolName)) {
+				throw new InvalidOperationException ($"Unsupported NDK tool '{kind}'");
+			}
+
+			return toolName;
+		}
+
+		protected static string GetArchTriple (AndroidTargetArch arch)
+		{
+			if (archTriples.TryGetValue (arch, out string? triple) && !String.IsNullOrEmpty (triple)) {
+				return triple;
+			}
+
+			throw new InvalidOperationException ($"Unsupported NDK architecture '{arch}'");
+		}
+
+		protected static string GetPlatformArch (AndroidTargetArch arch)
+		{
+			if (archPlatforms.TryGetValue (arch, out string? name) && !String.IsNullOrEmpty (name)) {
+				return name;
+			}
+
+			throw new InvalidOperationException ($"Unsupported NDK architecture '{arch}'");
+		}
+
+		protected string EnsureDirectoryExists (string path)
+		{
+			if (Directory.Exists (path)) {
+				return path;
+			}
+
+			throw new InvalidOperationException ($"Required directory '{path}' not found");
+		}
+
+		// This is an ugly compromise to support unified headers with minimum code duplication, because C# doesn't
+		// support multiple inheritance :(
+		//
+		// Unified headers are supported by both clang and non-clang NDKs, so the clang+unified headers NDK class
+		// (NdkToolsNoClangWithUnifiedHeaders) would have to derive from two classes - one to implement the "with clang"
+		// option and another to implement the "with unified headers" option.
+		protected string GetUnifiedHeadersDirPath (string androidNdkPath)
+		{
+			return EnsureDirectoryExists (MakeUnifiedHeadersDirPath (androidNdkPath));
+		}
+
+		protected virtual string MakeUnifiedHeadersDirPath (string androidNdkPath)
+		{
+			return Path.Combine (androidNdkPath, "sysroot", "usr", "include");
+		}
+
+		protected string UnifiedHeaders_GetAsmIncludeDirPath (AndroidTargetArch arch, int apiLevel)
+		{
+			return Path.Combine (GetUnifiedHeadersDirPath (NdkRootDirectory), GetArchTriple (arch));
+		}
+
+		protected string UnifiedHeaders_GetPlatformIncludeDirPath (AndroidTargetArch arch, int apiLevel)
+		{
+			return GetUnifiedHeadersDirPath (NdkRootDirectory);
+		}
+
+		const string platformLinux64 = "linux-x86_64";
+                const string platformLinux32 = "linux-x86";
```

Comment:
Indentation starting here is wrong, and appears to be using spaces instead of tabs.